nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
+    Targeting, // The targeting has changed for an experiment.
+}
+
 // Every experiment has an ExperimentEnrollment, even when we aren't enrolled.
 #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
 pub struct ExperimentEnrollment {
     pub slug: String,
     pub status: EnrollmentStatus,
 }
 
+pub struct EnrollmentChangeEvent {
+    pub experiment_slug: String,
+    pub branch_slug: String,
+    pub enrollment_id: String,
+    pub reason: Option<String>,
+    pub change: EnrollmentChangeEventType,
+}
+
+impl EnrollmentChangeEvent {
+    pub(crate) fn new(
+        slug: &str,
+        enrollment_id: &str,
+        branch: &str,
+        reason: Option<&str>,
+        change: EnrollmentChangeEventType,
+    ) -> Self {
+        Self {
+            experiment_slug: slug.to_owned(),
+            branch_slug: branch.to_owned(),
+            reason: reason.map(|s| s.to_owned()),
+            enrollment_id: enrollment_id.to_owned(),
+            change,
+        }
+    }
+}
+
+pub enum EnrollmentChangeEventType {
+    Enrollment,
+    Disqualification,
+    Unenrollment,
+}
+
+impl ExperimentEnrollment {
+    fn from_new_experiment(
+        is_user_participating: bool,
+        nimbus_id: &Uuid,
+        available_randomization_units: &AvailableRandomizationUnits,
+        app_context: &AppContext,
+        experiment: &Experiment,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Self> {
+        Ok(if !is_user_participating {
+            Self {
+                slug: experiment.slug.clone(),
+                status: EnrollmentStatus::NotEnrolled {
+                    reason: NotEnrolledReason::OptOut,
+                },
+            }
+        } else {
+            let enrollment = evaluate_enrollment(
+                nimbus_id,
+                available_randomization_units,
+                app_context,
+                experiment,
+            )?;
+            log::debug!(
+                "Experiment '{}' is new - enrollment status is {:?}",
+                &enrollment.slug,
+                &enrollment
+            );
+            if let EnrollmentStatus::Enrolled {
+                branch,
+                enrollment_id,
+                ..
+            } = &enrollment.status
+            {
+                enrollment_events.push(EnrollmentChangeEvent::new(
+                    &enrollment.slug,
+                    enrollment_id,
+                    branch,
+                    None,
+                    EnrollmentChangeEventType::Enrollment,
+                ));
+            };
+            enrollment
+        })
+    }
+    fn on_experiment_updated(
+        &self,
+        is_user_participating: bool,
+        nimbus_id: &Uuid,
+        available_randomization_units: &AvailableRandomizationUnits,
+        app_context: &AppContext,
+        updated_experiment: &Experiment,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Option<Self>> {
```

Comment:
If I'm reading the update code correctly, returning `None` here will result in the experiment not existing in the db?